File: 141. Linked List Cycle.java
import java.util.Set;
import java.util.HashSet;

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}


class Solution {

    private Set<ListNode> set = new HashSet<>();

    public boolean hasCycle(ListNode head) {

        if(head==null){
            return  false;
        }
        else if (head.next == null){
            return false;
        }

        ListNode slow = head;
        ListNode fast = head.next;

        while(true){

            if(slow == fast){
                return true;
            }
            if(fast.next == null || fast.next.next == null){
                return  false;
            }

            slow = slow.next;
            fast = fast.next.next;

            
        }
  }
}

File: 141. Linked List Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:


    def hasCycle(self, head: Optional[ListNode]) -> bool:

        hashset = set()

        while head != None:
            if head in hashset: return True
            else: hashset.add(head)
            head = head.next

        return False

    def constantMemoryhasCycle(self, head: Optional[ListNode]) -> bool:

        slow = fast = head 

        while fast != None and fast.next != None:
            slow = slow.next
            fast = (fast.next).next
            if slow == fast: return True

        return False
